packages/analyzer/src/javascript/esutil.ts

```diff
@@ -502,6 +502,32 @@ export function getMethods(node: babel.Node, document: JavaScriptDocument):
   return methods;
 }
 
+export function getConstructorMethod(
+    astNode: babel.Node, document: JavaScriptDocument): ScannedMethod|
+    undefined {
+  if (!babel.isClass(astNode)) {
+    return;
+  }
+
+  const statement = getConstructorClassMethod(astNode);
+
+  if (statement) {
+    const method = toScannedMethod(
+        statement, document.sourceRangeForNode(statement)!, document);
+    if (astNode.id) {
+      method.return = {...method.return, type: astNode.id.name};
```

Comment:
Surprisingly enough, a constructor can return something other than the class:

```javascript
class Bar {
}
class Foo {
  constructor() {
    return new Bar();
  }
}
console.log((new Foo()).constructor.name);
```

So maybe we should either not try to infer the `return` type or just defer to jsdoc if it states the return type. I think we've been leaning towards just not inferring stuff. I think the code we'll generate for declared TypeScript classes doesn't want a return type from the constructor anyways.

Reply:
Yes I was aware of that, so that's why I left it in but was not sure. So if I understand correctly you would just do nothing in this case and remove the return value all together? That would simplify the TypeScript case indeed.

Reply:
Two things: 

 - `method.return` may have a type from jsdoc, and this code will overwrite that if astNode.id exists. Documentation should take priority over inference.
 - Maybe we shouldn't infer the type at all for constructors, because we want to distinguish the cases where we have strong information from jsdoc vs the case where we're assuming that the constructor is a normal constructor and constructs the class, and just leaving the type off for the second case seems like the best option.